cmd/kubectl-gadget/utils/trace.go

```diff
@@ -464,61 +467,102 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
 	_, err = untilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
 		// This function will be executed until:
-		// 1. The number of watched traces equals the number of traces, i.e. we
-		// dealt with the traces which interest us.
+		// 1. The number of watched traces equals the number of traces to watch,
+		// i.e. we dealt with the traces which interest us.
 		// 2. Or it returns an error.
 		// 3. Or time out is fired.
 		// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
 		// with.
+		trace := event.Object.(*gadgetv1alpha1.Trace)
 
-		// Deal particularly with error.
-		if event.Type == watch.Error {
-			return false, err
-		}
-
-		// We are only interested in Added and Modified event, as we want
-		// Status.State value to change.
-		// More particularly, we monitor Added event for gadget like dns and
-		// Modified for gadget like seccompadvisor.
-		if event.Type != watch.Added && event.Type != watch.Modified {
+		// If the trace is already in the map, it means it already satisfies
+		// conditionFunction, so we can skip this event.
+		if _, present := watchedTraces[trace.ObjectMeta.Name]; present {
 			return false, nil
 		}
 
-		if event.Type == watch.Added {
-			// The API adds fake watch.Added events, so we use them to count the
-			// number of trace:
-			// To establish initial state, the watch begins with synthetic "Added"
-			// events of all resources instances that exist at the starting resource
-			// version
-			tracesNumber++
+		switch event.Type {
+		case watch.Deleted:
+			// If for some strange reasons (e.g. users deleted a trace during this
+			// operation) a trace is deleted, we need to take care of this by
+			// decrementing the tracesToWatch.
+			// Otherwise we would still wait for the old number and we would timeout.
+			tracesToWatch--
+
+			// We also remove it from the map to avoid returning a deleted trace.
+			delete(watchedTraces, trace.ObjectMeta.Name)
+		case watch.Modified:
+			// We will deal with this type of event below
+		case watch.Error:
+			// Deal particularly with error.
+			return false, err
+		case watch.Added:
+			// createTraces() creates traces synchronously.
+			// So, if a watch.Added event occurs it means there is a problem (e.g. the
+			// user creates a trace by snooping on the traceID of existing traces).
+			return false, fmt.Errorf("No traces with the given traceID (%s) should be created", traceID)
+		default:
+			// We are not interested in other event types.
+			return false, nil
 		}
 
-		trace := event.Object.(*gadgetv1alpha1.Trace)
-
 		if trace.Status.OperationWarning != "" {
+			// The trace can have a warning but satisfies conditionFunction.
+			// So, we do not add it to the map here.
 			nodeWarnings[trace.Spec.Node] = trace.Status.OperationWarning
 		}
 
 		if trace.Status.OperationError != "" {
-			watchedTracesNumber++
+			watchedTraces[trace.ObjectMeta.Name] = *trace
```

Comment:
Does it mean we can return traces with errors?

Reply:
It seems...
I will filter during the gather loop at the end of the function.

Reply:
It seems...
I will filter during the gather loop at the end of the function.

Reply:
Thinking to it I am not sure what is the good behavior here.
Like the above question, we can either return with the succeeded traces or we can decide to timeout on purpose because one trace is in error and does not satisfy `conditionFunction()`.

Reply:
IMO, we should not time out if there are traces that returned an error but `waitForCondition()` should return **only** successful traces. 

I think we can follow Mauro's suggestion [here](https://github.com/kinvolk/inspektor-gadget/pull/535#discussion_r815025457) and then filter out the traces with error.

Reply:
I changed the thing a little bit.
There are one map for traces which satisfy the `conditionFunction()` and another for traces which have `trace.Status.Operationerror != ""`.